maven/maven.bzl

```diff
@@ -207,18 +216,51 @@ def _check_for_duplicates(artifact_specs):
         elif sets.pop(versions).endswith("-SNAPSHOT"):
             fail("Snapshot versions are not supported in maven_artifacts.bzl.  Please fix %s to a pinned version." % artifact);
 
+_valid_artifact_spec_properties = sets.new("sha256", "insecure")
+def _unsupported_keys(keys_list):
+    return sets.difference(_valid_artifact_spec_properties, sets.copy_of(keys_list))
+
+def _fix_string_booleans(value):
+    if type(value) == type(""):
+        if value != "True" or value != "true":
+            return False
+    return bool(value)
+
 # If artifact/sha pair has missing sha hashes, reject it.
-def _validate_no_insecure_artifacts(artifact_specs = {}):
-    insecure_artifacts = {}
-    for spec, sha in artifact_specs.items():
-        if not bool(sha):
-            insecure_artifacts += { spec : sha }
-    if bool(insecure_artifacts):
-        fail("\n%s %s [\n%s]" % (
-             "These artifacts were specified without sha256 hashes.",
-             "Either add hashes or move to insecure_artifacts:",
-             "".join(sorted(["    \"%s\",\n" % x for x in insecure_artifacts.keys()]))
-        ))
+def _validate_artifacts(artifact_definitions = {}):
+    errors = []
+    if not bool(artifacts):
+        errors += ["At least one artifact must be specified."]
+    for spec, properties in artifact_definitions.items():
+        if type(properties) != type({}):
+            errors += ["""Artifact %s has an invalid property dictionary. Should not be a %s""", (spec, type(properties))]
+        unsupported_keys = _unsupported_keys(properties.keys())
+        if bool(unsupported_keys):
+            errors += ["""Artifact %s has unsupported property keys: %s. Only %s are supported""" % (
+                spec, list(unsupported_keys), list(_valid_artifact_spec_properties)
+            )]
+        artifact = artifacts.parse_spec(spec) # Basic sanity check.
+        if not bool(artifact.version):
+            errors += ["""Artifact "%s" missing version""" % spec]
+        if not properties.get("sha256", None) and not _fix_string_booleans(properties.get("insecure", False)):
+            errors += ["""Artifact "%s" is mising a sha256. Either supply it or mark it "insecure".""" % spec]
+    if bool(errors):
+        fail("Errors found:\n    %s" % "\n    ".join(errors))
+
+# Pre-process the artifact_declarations dictionary and handle older APIs that are now deprecated, until we
+# delete them.
+def _handle_legacy_specifications(artifact_declarations, insecure_artifacts):
+    # Legacy deprecated feature backwards compatibility
+    for spec in insecure_artifacts:
+        print(_INSECURE_DEPRECATION_WARNING % spec)
```

Comment:
Curious why these two warnings were broken out into constants but the ones in `_validate_artifacts` weren't.

Reply:
Good question. Inconsistency on my part.  I'll fix.

Reply:
Oh, I remember.  I moved the constants up because they're multiline.  The other ones fit inline.

Reply:
I think I'm going to leave the validation ones in the method, since you can read it all at one place.  the others are warnings which should all go away anyway once I remove the old property.